File: ruby/dialog/modesheet.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'cocoasheet'

class ModeSheet < CocoaSheet
  attr_accessor :mode, :uid, :cid
  ib_mapped_outlet :sCheck, :pCheck, :nCheck, :tCheck, :tCheck, :iCheck, :mCheck, :aCheck, :rCheck
  ib_outlet :kCheck, :lCheck, :password, :limit
  first_responder :sCheck
  buttons :Ok, :Cancel
  
  def startup(chname, mode)
    @mode = mode.dup
    load
    update
    
    case chname.first_char
    when '!'
      @aCheck.setEnabled(true)
      @rCheck.setEnabled(true)
    when '&'
      @aCheck.setEnabled(true)
      @rCheck.setEnabled(false)
    else
      @aCheck.setEnabled(false)
      @rCheck.setEnabled(false)
    end
  end
  
  def shutdown(result)
    if result == :ok
      save
      fire_event('onOk', @mode)
    else
      fire_event('onCancel')
    end
  end

  def load
    load_mapped_outlets(@mode)
    @kCheck.setState(!@mode.k.empty? ? 1 : 0)
    @lCheck.setState(@mode.l > 0 ? 1 : 0)
    @password.setStringValue(@mode.k)
    @limit.setIntValue(@mode.l)
  end
  
  def save
    save_mapped_outlets(@mode)
    if @kCheck.state == 1 && !@password.stringValue.to_s.empty?
      @mode.k = @password.stringValue.to_s
    else
      @mode.k = ''
    end
    if @lCheck.state == 1 && !@limit.stringValue.to_s.empty?
      @mode.l = @limit.stringValue.to_s.to_i
    else
      @mode.l = 0
    end
  end
  
  def update
    @password.setEnabled(@kCheck.state == 1)
    @limit.setEnabled(@lCheck.state == 1)
  end
  
  def onChangeChecks(sender)
    if @sCheck.state == 1 && @pCheck.state == 1
      case sender.__ocid__
      when @sCheck.__ocid__; @pCheck.setState(0)
      when @pCheck.__ocid__; @sCheck.setState(0)
      end
    end
    update
  end
end

File: objc/Dialogs/ModeSheet.m
// Created by Satoshi Nakagawa.
// You can redistribute it and/or modify it under the Ruby's license or the GPL2.

#import "ModeSheet.h"


@interface ModeSheet (Private)
- (void)updateTextFields;
@end


@implementation ModeSheet

@synthesize mode;
@synthesize channelName;
@synthesize uid;
@synthesize cid;

- (id)init
{
	if (self = [super init]) {
		[NSBundle loadNibNamed:@"ModeSheet" owner:self];
	}
	return self;
}

- (void)dealloc
{
	[mode release];
	[channelName release];
	[super dealloc];
}

- (void)start
{
	[sCheck setState:mode.s];
	[pCheck setState:mode.p];
	[nCheck setState:mode.n];
	[tCheck setState:mode.t];
	[iCheck setState:mode.i];
	[mCheck setState:mode.m];
	[aCheck setState:mode.a];
	[rCheck setState:mode.r];
	[kCheck setState:mode.k.length > 0];
	[lCheck setState:mode.l > 0];
	
	[kText setStringValue:mode.k ?: @""];
	[lText setStringValue:[NSString stringWithFormat:@"%d", mode.l]];
	
	[self updateTextFields];

	if ([channelName hasPrefix:@"!"]) {
		[aCheck setEnabled:YES];
		[rCheck setEnabled:YES];
	}
	else if ([channelName hasPrefix:@"&"]) {
		[aCheck setEnabled:YES];
		[rCheck setEnabled:NO];
	}
	else {
		[aCheck setEnabled:NO];
		[rCheck setEnabled:NO];
	}
	
	[sheet makeFirstResponder:sCheck];
	[self startSheet];
}

- (void)updateTextFields
{
	[kText setEnabled:kCheck.state == NSOnState];
	[lText setEnabled:lCheck.state == NSOnState];
}

- (void)onChangeCheck:(id)sender
{
	[self updateTextFields];
	
	if ([sCheck state] == NSOnState && [pCheck state] == NSOnState) {
		if (sender == sCheck) {
			[pCheck setState:NSOffState];
		}
		else {
			[sCheck setState:NSOffState];
		}
	}
}

- (void)ok:(id)sender
{
	mode.s = [sCheck state] == NSOnState;
	mode.p = [pCheck state] == NSOnState;
	mode.n = [nCheck state] == NSOnState;
	mode.t = [tCheck state] == NSOnState;
	mode.i = [iCheck state] == NSOnState;
	mode.m = [mCheck state] == NSOnState;
	mode.a = [aCheck state] == NSOnState;
	mode.r = [rCheck state] == NSOnState;
	
	if ([kCheck state] == NSOnState) {
		mode.k = [kText stringValue];
	}
	else {
		mode.k = @"";
	}
	
	if ([lCheck state] == NSOnState) {
		mode.l = [[lText stringValue] intValue];
	}
	else {
		mode.l = 0;
	}
	
	if ([delegate respondsToSelector:@selector(modeSheetOnOK:)]) {
		[delegate modeSheetOnOK:self];
	}
	
	[super ok:sender];
}

#pragma mark -
#pragma mark NSWindow Delegate

- (void)windowWillClose:(NSNotification*)note
{
	if ([delegate respondsToSelector:@selector(modeSheetWillClose:)]) {
		[delegate modeSheetWillClose:self];
	}
}

@end
